### DISCUSSION TITLE
How to produce single exe file after push for unpackaged app ?

### DISCUSSION THREAD

**Discussion by IvanDmitriev1** [microsoft/WindowsAppSDK#2755] [Q&A] [answered]
I used windows app sdk version 1.1.2

With these push settings:
![Screenshot_1](https://user-images.githubusercontent.com/42055372/180037082-f63194d2-28d9-4a58-ab2a-6017ab71658f.png)

And what i get with `produce single file` option:
![Screenshot_2](https://user-images.githubusercontent.com/42055372/180037387-b3e5aa28-9b51-4365-8adc-bc77e8cb4fca.png)

~~Why `Microsoft.WindowsAppRuntime.Bootstrap.dll` not shipped with windows app runtime or in part of exe ?~~
Why `resources.pri` is not in exe itself ?

**Comment by ghost1372** [comment]
Single exe file is not working dont use it

**Reply by IvanDmitriev1** [reply]
For me its working.

**Reply by ghost1372** [reply]
> For me its working.

Put a InfoBar element
Now its not working😅

**Comment by DrusTheAxe** [comment]
>Why Microsoft.WindowsAppRuntime.Bootstrap.dll not shipped with windows app runtime

It's a redistributable. Unpackaged apps need to (successfully) call `MddBootstrapInitialize()` before you can use the runtime so it can't go into the runtime (the usual chicken/egg problem).

>or in part of exe ?

It could be provided as a static library but that brings a ton of complications to the table starting with "Assumes you're using the same compiler version and settings" e.g. no help C# considering it's not implemented in C#. Or Rust. Or... It could be provided as source code but it's non-trivial and again assumes you're using C++...

I thought there was more to make 'single file' work than this. @mikebattista ?

**Reply by IvanDmitriev1** [reply]
But what about Microsoft.WindowsAppRuntime.Bootstrap.dll save it somewhere in a known place along with installing the win app runtime, but why isn't it done this way?

And after all, Microsoft.WindowsAppRuntime.Bootstrap.dll it seems to depend on the win app runtime so that there is some sense to download it together with the runtime.

**Comment by Balkoth** [comment, marked as answer]
You will need to include `<IncludeNativeLibrariesForSelfExtract>true</IncludeNativeLibrariesForSelfExtract>` but resources.pri will still be separate. 

Honestly this whole WindowsAppSdk thing is not production ready. Even the beta tag would be a bold statement. The quality bar for the release version is exceptionally low for a microsoft product. WPF had its quirks when it first came out, but this whole stack is extremely frustrating to work with.

The debugging experience in C# turns out to be a nightmare when you leave your hello world program.

Every new version includes new regressions and bugs to basic functionality. Bugs like drag and drop crashing the app are there through numerous releases. Bugs like scrolling a listview crashing the app show that there is no testing in place even for the most elementary functions of the controls.

This all has been mentioned and discussed in numerous community calls in which they always promise to do better. Nothing groundbreaking has happened so far. The microsoft-ui-xaml and this repository keep racking up issues in the thousands.

I can't believe that the best strategy in 2022, if you want to develop a windows app, with a nice GUI, is to stick to plain old windows forms and look for controls from a professional vendor like DevExpress, etc. You get designer support (yes WinUI3 is lacking a real graphical designer), a proper debugging experience (did i mention the WinUI3 debugging experience is complete crap) and with all this tried and tested tools that just work.

P.S.: I am mad and im venting here 😡

**Reply by WaGi-Coding** [reply]
I totally feel you ^^ I stick to boring Winforms, just because it works and is easy to publish smaller weekend projects. With that WinUI stuff, you have more trouble with getting it to work in first place than with the actual functionality of your program

**Reply by mbodm** [reply]
> I totally feel you ^^ I stick to boring Winforms, just because it works and is easy to publish smaller weekend projects. With that WinUI stuff, you have more trouble with getting it to work in first place than with the actual functionality of your program

Sooo true. Sad to say so :(

**Comment by mbodm** [comment]
It´s May 2023 and i still would sign what @Balkoth said above. That whole beast is just a nightmare of its own. Even from a C# perspective. I don´t wanna even start talking about the C++ quirks i stepped over, when tested the whole stuff, for more than 40 hours.

But the most disappointing aspect to me was the output size. Coming from 144 MB sized single EXE files (self-contained WinForms/WPF applications), i decided to check the size of a WinUI3 Hello-World app. Which ended up with a 119 MB EXE + 2 external files, which not really can be compiled into the EXE (see above posts). So, for C# there is still the framework-dependent way. Doing this in C# and WinForms/WPF is just a walk in the park. Everything works out of the box, without any quirks. Create your 145 KB  framework-dependent WinForms/WPF EXE, make sure your clients have .NET 6 (or .NET 5, or .NET 7, or whatever) installed and you will sleep like a baby. Do the same thing with Windows App SDK and WinUI3 and see what happens... 😉 

Besides the size, what i experienced working with Windows App SDK and WinUI3 for 40-50 hours in May 2023, was all in all one of the most horrible experiences i had in the last 10 years, when it comes down to Desktop UI development. At its best i would consider this a late beta. On good days. Not what i normally expect to see from MS. Sad to say. 😞

**Reply by IvanDmitriev1** [reply]
Hi I'm working on [fluent UI for wpf](https://github.com/lepoco/wpfui), maybe you'll like it.
